### PR TITLE
Decharging material density and recipe fixes

### LINKED ISSUE #87688: Decharging has significantly changed the density and unit weight/volume of many items
### Describe the bug

Several items have recently been moved from charges to being treated as individual items (#87323, #87332, #87180, #87008, #87006, #87000). The semantics for charged items is that the `volume` field is the volume for a full-stack, while the `weight` field is the unit weight. So, when decharging an item the volume must be reduced by `1/count` in order to preserve the original unit volume. This is complicated by the 1mL lower limit on volume, so in cases where `old_stack_volume/count < 1mL`, we have been setting the new unit volume to 1mL and adjusting the weight to preserve the original density. In this case, crafting recipes and item spawning would need to be adjusted for the change in the actual amount of material per unit. These things have not been done consistently so far, so there are items which now have significantly different unit weight/volume and density, and/or for which crafting recipes have not been adjusted.

There are several existing issues for this type of problem, but I figure it's easier to try to resolve all the cases systematically in one go, so I'm making a centralized issue for it.

I wrote a script to calculate weight/volume/density changes, and ran against the current state in 965b136abf1e7599dd2cc46a583c9de6630c8238. Bold items are changes are >10%.

| id | unit weight (mg) | unit volume (mL) | density (mg&#x2f;mL) |
| --- | --- | --- | --- |
| chem_black_powder | 151.0 → 151.0 (1.0x) | **0.09 → 89.0 (1000.0x)** | **1696.63 → 1.7 (0.001x)** |
| chem_rocket_fuel | 70000.0 → 70000.0 (1.0x) | **25.0 → 250.0 (10.0x)** | **2800.0 → 280.0 (0.1x)** |
| welding_rod_steel | 2777.0 → 2777.0 (1.0x) | **1.2 → 1.0 (0.8333x)** | **2314.17 → 2777.0 (1.2x)** |
| welding_rod_alloy | 1429.0 → 1429.0 (1.0x) | **1.2 → 1.0 (0.8333x)** | **1190.83 → 1429.0 (1.2x)** |
| welding_wire_steel | 2777.0 → 2777.0 (1.0x) | **0.72 → 1.0 (1.3846x)** | **3845.08 → 2777.0 (0.7222x)** |
| welding_wire_alloy | 1429.0 → 1429.0 (1.0x) | **1.39 → 1.0 (0.72x)** | **1028.88 → 1429.0 (1.3889x)** |
| brazing_rod_bronze | 3200.0 → 3200.0 (1.0x) | **1.6 → 2.0 (1.25x)** | **2000.0 → 1600.0 (0.8x)** |
| brazing_rod_alloy | 1750.0 → 1750.0 (1.0x) | **1.5 → 2.0 (1.3333x)** | **1166.67 → 875.0 (0.75x)** |
| aluminum_foil | 135.0 → 135.0 (1.0x) | **0.05 → 10.0 (200.0x)** | **2700.0 → 13.5 (0.005x)** |
| copper_wire | **4747.0 → 9000.0 (1.8959x)** | **0.55 → 1.0 (1.8182x)** | 8630.91 → 9000.0 (1.0428x) |
| down_feather | **10.0 → 1.0 (0.1x)** | **12.5 → 1.0 (0.08x)** | **0.8 → 1.0 (1.25x)** |
| lc_chain_link | 320.0 → 320.0 (1.0x) | **0.59 → 250.0 (425.0x)** | **544.0 → 1.28 (0.0024x)** |
| chem_potassium_alum | 1725.0 → 1725.0 (1.0x) | 1.0 → 1.0 (1.0x) | 1725.0 → 1725.0 (1.0x) |
| dye_powder | 1540.0 → 1540.0 (1.0x) | 1.0 → 1.0 (1.0x) | 1540.0 → 1540.0 (1.0x) |
| chem_calcium_chloride | 2150.0 → 2150.0 (1.0x) | 1.0 → 1.0 (1.0x) | 2150.0 → 2150.0 (1.0x) |
| chem_washing_soda | 2540.0 → 2540.0 (1.0x) | 1.0 → 1.0 (1.0x) | 2540.0 → 2540.0 (1.0x) |
| chem_baking_soda | 2200.0 → 2200.0 (1.0x) | 1.0 → 1.0 (1.0x) | 2200.0 → 2200.0 (1.0x) |
| chem_mercury_fulminate | 442.0 → 442.0 (1.0x) | **0.1 → 250.0 (2500.0x)** | **4420.0 → 1.77 (0.0004x)** |
| chem_lead_azide | 471.0 → 471.0 (1.0x) | **0.1 → 250.0 (2500.0x)** | **4710.0 → 1.88 (0.0004x)** |
| solder_wire | 2000.0 → 2000.0 (1.0x) | **0.25 → 1.0 (4.0x)** | **8000.0 → 2000.0 (0.25x)** |
| cable | **5413.0 → 10000.0 (1.8474x)** | **0.78 → 2.0 (2.5806x)** | **6984.52 → 5000.0 (0.7159x)** |
| feather | **1000.0 → 10.0 (0.01x)** | **12.5 → 1.0 (0.08x)** | **80.0 → 10.0 (0.125x)** |
| gunpowder | 100.0 → 100.0 (1.0x) | **0.06 → 250.0 (4540.0x)** | **1816.0 → 0.4 (0.0002x)** |
| oxy_powder | 1000.0 → 1000.0 (1.0x) | **1.25 → 1.0 (0.8x)** | **800.0 → 1000.0 (1.25x)** |
| lye_powder | 2662.0 → 2662.0 (1.0x) | **1.25 → 1.0 (0.8x)** | **2129.6 → 2662.0 (1.25x)** |
| shotgun_primer | 1000.0 → 1000.0 (1.0x) | **1.25 → 1.0 (0.8x)** | **800.0 → 1000.0 (1.25x)** |
| smpistol_primer | 1000.0 → 1000.0 (1.0x) | **1.25 → 1.0 (0.8x)** | **800.0 → 1000.0 (1.25x)** |
| lgpistol_primer | 1000.0 → 1000.0 (1.0x) | **1.25 → 1.0 (0.8x)** | **800.0 → 1000.0 (1.25x)** |
| smrifle_primer | 1000.0 → 1000.0 (1.0x) | **1.25 → 1.0 (0.8x)** | **800.0 → 1000.0 (1.25x)** |
| lgrifle_primer | 1000.0 → 1000.0 (1.0x) | **1.25 → 1.0 (0.8x)** | **800.0 → 1000.0 (1.25x)** |
| spent_shotgun_primer | 1000.0 → 1000.0 (1.0x) | **1.25 → 1.0 (0.8x)** | **800.0 → 1000.0 (1.25x)** |
| spent_smpistol_primer | 1000.0 → 1000.0 (1.0x) | **1.25 → 1.0 (0.8x)** | **800.0 → 1000.0 (1.25x)** |
| spent_lgpistol_primer | 1000.0 → 1000.0 (1.0x) | **1.25 → 1.0 (0.8x)** | **800.0 → 1000.0 (1.25x)** |
| spent_smrifle_primer | 1000.0 → 1000.0 (1.0x) | **1.25 → 1.0 (0.8x)** | **800.0 → 1000.0 (1.25x)** |
| spent_lgrifle_primer | 1000.0 → 1000.0 (1.0x) | **1.25 → 1.0 (0.8x)** | **800.0 → 1000.0 (1.25x)** |
| rubber_slug | 2000.0 → 2000.0 (1.0x) | **10.0 → 250.0 (25.0x)** | **200.0 → 8.0 (0.04x)** |
| copper | 5000.0 → 5000.0 (1.0x) | **0.64 → 1.0 (1.5625x)** | **7812.5 → 5000.0 (0.64x)** |
| lamellae_lc | 14000.0 → 14000.0 (1.0x) | **5.56 → 100.0 (18.0x)** | **2520.0 → 140.0 (0.0556x)** |
| lamellae_ch | 14000.0 → 14000.0 (1.0x) | **5.56 → 100.0 (18.0x)** | **2520.0 → 140.0 (0.0556x)** |
| button_plastic | 1000.0 → 1000.0 (1.0x) | **1.0 → 20.0 (20.0x)** | **1000.0 → 50.0 (0.05x)** |
| incendiary | 2000.0 → 2000.0 (1.0x) | **2.5 → 500.0 (200.0x)** | **800.0 → 4.0 (0.005x)** |
| fuse | 6000.0 → 6000.0 (1.0x) | **1.6 → 80.0 (50.0x)** | **3750.0 → 75.0 (0.02x)** |
| slag | 10000.0 → 10000.0 (1.0x) | **2.5 → 3.0 (1.2x)** | **4000.0 → 3333.33 (0.8333x)** |
| graphite | **40.0 → 2000.0 (50.0x)** | **0.02 → 1.0 (50.0x)** | 2000.0 → 2000.0 (1.0x) |
| chem_sulphur | **320.0 → 2000.0 (6.25x)** | **0.16 → 1.0 (6.25x)** | 2000.0 → 2000.0 (1.0x) |
| chem_borax | 3810.0 → 3460.0 (0.9081x) | 2.2 → 2.0 (0.9091x) | 1731.82 → 1730.0 (0.999x) |
| shredded_rubber | 6000.0 → 6000.0 (1.0x) | 5.0 → 5.0 (1.0x) | 1200.0 → 1200.0 (1.0x) |
| material_limestone | 14000.0 → 14000.0 (1.0x) | 5.0 → 5.0 (1.0x) | 2800.0 → 2800.0 (1.0x) |
| chem_slaked_lime | 2211.0 → 2211.0 (1.0x) | 1.0 → 1.0 (1.0x) | 2211.0 → 2211.0 (1.0x) |
| chem_aluminium_powder | 1760.0 → 1760.0 (1.0x) | **2.5 → 3.0 (1.2x)** | **704.0 → 586.67 (0.8333x)** |
| chem_zinc_powder | **654.0 → 7140.0 (10.9174x)** | **0.09 → 1.0 (11.1111x)** | 7266.67 → 7140.0 (0.9826x) |
| chem_zinc_oxide | **814.0 → 5606.0 (6.887x)** | **0.15 → 1.0 (6.6667x)** | 5426.67 → 5606.0 (1.033x) |
| chem_manganese_dioxide | **869.0 → 5026.0 (5.7837x)** | **0.17 → 1.0 (5.8824x)** | 5111.76 → 5026.0 (0.9832x) |
| cathod_mix | 500.0 → 500.0 (1.0x) | **0.15 → 1.0 (6.6225x)** | **3311.26 → 500.0 (0.151x)** |
| chem_ferric_chloride | 2900.0 → 2900.0 (1.0x) | 1.0 → 1.0 (1.0x) | 2900.0 → 2900.0 (1.0x) |
| chem_antimony_trichloride | 3140.0 → 3140.0 (1.0x) | 1.0 → 1.0 (1.0x) | 3140.0 → 3140.0 (1.0x) |
| chem_potassium_chloride | **746.0 → 1984.0 (2.6595x)** | **0.38 → 1.0 (2.6316x)** | 1963.16 → 1984.0 (1.0106x) |
| chem_potassium_hydroxide | **561.0 → 2120.0 (3.779x)** | **0.26 → 1.0 (3.8462x)** | 2157.69 → 2120.0 (0.9825x) |
| chem_chromium_oxide | **1520.0 → 5220.0 (3.4342x)** | **0.29 → 1.0 (3.4483x)** | 5241.38 → 5220.0 (0.9959x) |
| chem_carbide | **641.0 → 2220.0 (3.4633x)** | **0.29 → 1.0 (3.4483x)** | 2210.34 → 2220.0 (1.0044x) |
| chem_caco3 | **783.0 → 2710.0 (3.461x)** | **0.29 → 1.0 (3.4483x)** | 2700.0 → 2710.0 (1.0037x) |
| chem_ammonium_nitrate | 1000.0 → 1000.0 (1.0x) | **0.58 → 1.0 (1.7241x)** | **1724.14 → 1000.0 (0.58x)** |
| chem_ammonium_nitrate_pellets | 1000.0 → 1000.0 (1.0x) | **1.25 → 1.0 (0.8x)** | **800.0 → 1000.0 (1.25x)** |
| chem_saltpetre | 5275.0 → 5275.0 (1.0x) | **5.0 → 50.0 (10.0x)** | **1055.0 → 105.5 (0.1x)** |
| chem_aluminium_sulphate | 13350.0 → 13350.0 (1.0x) | 5.0 → 5.0 (1.0x) | 2670.0 → 2670.0 (1.0x) |
| chem_thermite | 10000.0 → 10000.0 (1.0x) | 5.0 → 5.0 (1.0x) | 2000.0 → 2000.0 (1.0x) |
| chem_anfo | 1000.0 → 1000.0 (1.0x) | **0.59 → 1.0 (1.7007x)** | **1700.68 → 1000.0 (0.588x)** |
| chem_rdx | **1000.0 → 1858.0 (1.858x)** | **0.54 → 1.0 (1.8519x)** | 1851.85 → 1858.0 (1.0033x) |
| chem_compositionb | **1000.0 → 1650.0 (1.65x)** | **0.6 → 1.0 (1.6667x)** | 1666.67 → 1650.0 (0.99x) |
| chem_hmtd | **1000.0 → 1570.0 (1.57x)** | **0.7 → 1.0 (1.4286x)** | 1428.57 → 1570.0 (1.099x) |
| chem_apex | **1000.0 → 500.0 (0.5x)** | **0.7 → 1.0 (1.4286x)** | **1428.57 → 500.0 (0.35x)** |
| detergent | 7500.0 → 7500.0 (1.0x) | 12.5 → 13.0 (1.04x) | 600.0 → 576.92 (0.9615x) |
| soap_flakes | 4000.0 → 4000.0 (1.0x) | 12.5 → 13.0 (1.04x) | 320.0 → 307.69 (0.9615x) |
| nanomaterial | 250000.0 → 250000.0 (1.0x) | 300.0 → 300.0 (1.0x) | 833.33 → 833.33 (1.0x) |
| ash | 1000.0 → 1000.0 (1.0x) | **1.64 → 2.0 (1.2195x)** | **609.76 → 500.0 (0.82x)** |
| corpse_ash | 1000.0 → 1000.0 (1.0x) | 1.04 → 1.0 (0.9615x) | 961.54 → 1000.0 (1.04x) |
| wooden_bead | 1000.0 → 1000.0 (1.0x) | **2.5 → 2.0 (0.8x)** | **400.0 → 500.0 (1.25x)** |
| rosin | 1000.0 → 1000.0 (1.0x) | 1.0 → 1.0 (1.0x) | 1000.0 → 1000.0 (1.0x) |
| iodine_crystal | 3500.0 → 3500.0 (1.0x) | 1.0 → 1.0 (1.0x) | 3500.0 → 3500.0 (1.0x) |
| scrap | 50000.0 → 50000.0 (1.0x) | **125.0 → 31.0 (0.248x)** | **400.0 → 1612.9 (4.0323x)** |
| bismuth | **2000.0 → 10000.0 (5.0x)** | **0.5 → 1.0 (2.0x)** | **4000.0 → 10000.0 (2.5x)** |
| gold_small | **5000.0 → 19000.0 (3.8x)** | **0.57 → 1.0 (1.7544x)** | **8771.93 → 19000.0 (2.166x)** |
| platinum_small | **5000.0 → 21000.0 (4.2x)** | **0.52 → 1.0 (1.9231x)** | **9615.38 → 21000.0 (2.184x)** |
| zinc_metal | **1785.0 → 7140.0 (4.0x)** | **0.5 → 1.0 (2.0x)** | **3570.0 → 7140.0 (2.0x)** |
| lead | 5000.0 → 5000.0 (1.0x) | 0.97 → 1.0 (1.0309x) | 5154.64 → 5000.0 (0.97x) |
| magnesium | 1120.0 → 1120.0 (1.0x) | **2.5 → 2.0 (0.8x)** | **448.0 → 560.0 (1.25x)** |
| silver_small | 5000.0 → 5000.0 (1.0x) | 1.05 → 1.0 (0.9524x) | 4761.9 → 5000.0 (1.05x) |
| tin | 2000.0 → 2000.0 (1.0x) | **0.5 → 1.0 (2.0x)** | **4000.0 → 2000.0 (0.5x)** |

### Attach save file

N/A

### Steps to reproduce

N/A

### Expected behavior

N/A

### Screenshots

_No response_

### Versions and configuration

commit 965b136abf1e7599dd2cc46a583c9de6630c8238

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by azraneth**
#### Summary
Bugfixes "fixed material density & recipes post decharging"

#### Purpose of change
fixes #87688, A flare shouldn't disassemble into 200x its volume. 

#### Describe the solution
Changed density values of the outliers listed in #87688. Changes all related recipes by scaling components based on weight. A change to the gunpowder in plastic bottle item group and the welding spool container item was required due to container limitations. 

#### Describe alternatives you've considered

Also do sweeping changes to itemgroups. For example aluminum foil is an outlier of spawning in way too large quantities at the moment.
Change some bullet crafting recipes to have fewer edge cases like pistol rounds now requiring relatively ~3x more gunpowder to craft. 
Let the recipes break thermodynamics. 

#### Testing
Ran all json tests.  Some known bad uncrafts got fixed and some more were added.

#### Additional context
No mods were touched, so some followup would be needed to be done there, especially in cases of gunpowder or aluminum foil. Bullet crafting in general is really messed up right now, with almost none of the uncrafts preserving weight or volume.

### CODE REVIEW

File: data/json/items/ammo.json [hunk 1]
```diff
@@ -773,10 +773,10 @@
     "color": "dark_gray",
     "description": "Firearm-quality gunpowder, mixed without concern for composition, burn rate, or shape, making it thoroughly useless for reloading ammunition.  It has been slightly ground to make its properties more uniform.",
     "sealed": false,
-    "material": [ "powder" ],
+    "material": [ "dense_powder" ],
     "flags": [ "TINDER" ],
-    "volume": "250 ml",
-    "weight": "100 mg",
+    "volume": "1 ml",
+    "weight": "1816 mg",
```
Comment (RenechCDDA) [inline data/json/items/ammo.json:779]: You're making this 18x heavier while taking up 250x less volume?

You've also adjusted none of the gunpowder recipes, despite changing all the gunpowders. They're all based on 100mg units, but you've... made them 18x heavier.

I honestly do not understand many of the changes in this PR, as e.g. gunpowder is an item which specifically got *skipped*. How did you determine the values you've written here?

Comment (azraneth) [reply data/json/items/ammo.json:779]: According to #87008 gunpowder does not seem to have been skipped at all, unless I misunderstand that change (gunpowder no longer has a count). Pre that change the density of gunpowder was ~1.8g/ml and that is now again reflected in the gunpowder item. If I missed some gunpowder recipe I'll take another look when i have time, likely in a couple of days, but all consumers of gunpowder should be updated. Notably if gunpowder recipes were designed around 100ml units that will be a problem due to item minimum size.

All new numbers were taken straight out of the linked issue #87688 and should represent the pre-decharge values.

Comment (RenechCDDA) [reply data/json/items/ammo.json:779]: The count is on the next line. I can see it as I am typing this comment.

Image

We can't just assume that the table in the linked issue is accurate. There was no confirmation of the linked issue, no scrutiny, no discussion... it is liable to *just plain being wrong*. (Any claim that gunpowder does not have a count is, obviously, wrong.)

Even assuming you took values directly from the table, I still have to ask how you came up with the exact values and implemented them. How did you go from table data to doing the math and entering new values? It doesn't look like you've done this by hand.

Comment (azraneth) [reply data/json/items/ammo.json]: I... genuinely can't believe I missed the count line, thank you for catching that.

The density changes are actually bespoke, although the followup changes to recipes are indeed agentic, but double checked. My thoughts for setting the new values was:  
* try to minimize changes to items where possible
* try to match old densities by changing weight only to not mess up spawns
* where items had what looked like a bespoke change to weight (f.e gold being set to exactly 19g) instead change volume, where this clashed with density matching both were changed (bismuth) 
* where items obviously had charge volumes (feathers, gunpowder (woops), lamellae) change the volumes down to match weight

If you think I should have other focuses I'm definitely willing to change some of these, but these made sense to me. I'll double check each density for these materials but my intuition tells me most of them are correct, apart from possibly the rare metals.

What I didn't properly double check is the agent I told to scale the recipe values by mass, since I didn't think of the fact that the gunpowder recipes use other ids, but since the change to gunpowder is unneccessary it will be reverted as soon as I get back to my dev environment anyway.
 
After a double check of real-world densities and a reversion to the gunpowder stuff, are there any other blockers you'd like cleared?

**Comment by RenechCDDA** [comment]
[We cannot and will not accept anything produced with the assistance or use of LLMs, "AI", etc.](https://github.com/CleverRaven/Cataclysm-DDA/blob/cfe6af8c7357de5d4c23ea2eeac8beaf77db3e4c/CONTRIBUTING.md#licensing-and-authorship)

image

**Comment by azraneth** [comment]
Incredible timing, as I just now redid almost all of this by hand (admittedly the entering of the new numbers into recipes was machine assisted. I don't suppose you'd consider reopening this if I manually enter the same numbers into each of the ~500 recipes?
